Weather App (Workshop Bounty) (#1370)
hackclub/hackclub: workshops/weather_app/README.md

Before we start building our app let's see how the final version would look like:

![main](https://cloud-ktt2vamjn.vercel.app/0screencapture-file-home-gautamjajoo-desktop-card-index-html-2020-10-09-00_34_48.png)

Isn't it amazing? Get your cup of coffee and 20 minutes of your time to build an awesome weather app. 

The [Source Code](https://github.com/gautamjajoo/BasicWeather)  &  [Demo](https://WeatherApp.gautamjajoo.repl.co) are hyperlinked here.

## 0. What could you expect till the end of the workshop?

Till the end of the workshop you would have a good understanding of how to deal with APIs. You could also expand this knowledge to be
creative and implement new APIs in your projects.

Also, you would learn some nice `CSS` tricks for making beautiful card designs and basic elements/properties of `JS`.

## 1. The Prerequisites
Some basic understanding of `HTML,CSS and Javascipt/JSON`.

Don't be upset if you don't know any of the above, we will make sure you learn something new till the end of the workshop.

## 2. Setting up the project on Repl.it

> Repl.it is an powerful online IDE(integrated development environment) where you can edit, build and collabrate to make some beautiful projects.

Reviewer: ```suggestion
Repl.it is a powerful online IDE(integrated development environment) where you can edit, build, and collaborate to make some beautiful projects.
```suggestion
Repl.it is a powerful online IDE(integrated development environment) where you can edit, build, and collaborate to make some beautiful projects.

```
```

Make sure you have set up your account on [Repl](repl.it). Once you are finished setting up your account, create a [new repl project](https://repl.it/languages/html) with the above 
mentioned languages.

Three files each of ```HTML,CSS and JS``` would be created and name them as  `index.html`, `style.css` and `app.js` respectively.
After the above instructions your screen would look something like this:

![repl.it screen](https://cloud-o4m9cgxr8.vercel.app/0screenshot_from_2020-10-09_12-14-29.png)

> Once you have finsihed setting up the files, we can now move on to writing some code.

Reviewer: ```suggestion
Once you have finished setting up the files, we can now move on to writing some code.

```

```css
*{
  margin: 0;
  padding: 0;
}

body {
  background-color: #dfe7ee; // pale blue
  line-height: 1.5;
  font-size: 125%;
  display: flex;
}
```

All our body part is done, now we can start making the Card.

```css
.container {
  position: relative;
  background-color: #fff;
  padding: 0 1em 1em; 
  margin: 150px 500px 400px auto; 
  -webkit-filter: drop-shadow(0 1em 1em rgba(0, 0, 0, 0.1));
          filter: drop-shadow(0 1em 1em rgba(0, 0, 0, 0.1));
}
```
Here `position:relative` is used for the pseudo-element which we will create in the next step. 

`webkit-filter` with `drop-shadow` is also a very cool property which is used for giving a highlighted shadow behind the card.

For making a tilted card we use, 

```css
.container::before {
  content: '';
  padding-top: 5%;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 100%;
  background-image: linear-gradient(3deg, white calc(50% - 1px), transparent 50%);
}

```

`::before` creates a pseudo-element that is the first child of the selected element. It is used to insert something before the selected element i.e. `.container`. It also adds some cosmetic content to an element with the content property. 

> A pseudo class such as `::before` has a default `position` value of `inline`. Therefore we change the `position` to `absolute` as elements whose position value is `absolute` are taken out of flow and are `block-level` elements with a default display CSS property value of block.

Reviewer: ```suggestion
A pseudo-class such as `::before` has a default `position` value of `inline`. Therefore we change the `position` to `absolute` as elements whose position value is `absolute` are taken out of the flow and are `block-level` elements with a default display CSS property value of the block.

```

At this point the  css code would look something like [this](https://repl.it/@gautamjajoo/Weather#style.css).

## 5. Adding JS and learning how to work with APIs

### What are APIs and how do they work?

API stands for Application Programming Interface. An API is a messenger that delivers your request to the provider that you’re requesting it from and then delivers the response back to you. In layman terms, an API is like a waiter in a restaurant where you are the customer and chef is the provider. You order the food to the waiter and the waiter informs the chef and then serves the food back to the customer.

Developers use APIs to make their jobs more efficient by reusing code and only changing the part that is relevant to the process they want to improve.

For this project we would be using a weather API which is given by [OpenWeather](https://openweathermap.org/api). 

To start with, you will have to create an account on OpenWeather to generate an API key for yourself.

![login](https://cloud-dk4z6apbz.vercel.app/0login.gif)

After you have registered head over to the [API](https://openweathermap.org/api) section. 

> We would be using the `Current Weahter Data` API. Subscribe to the API and after that you will receive the key in your [profile](https://home.openweathermap.org/api_keys).

Reviewer: ```suggestion
We would be using the `Current Weather Data` API. Subscribe to the API and after that, you will receive the key in your [profile](https://home.openweathermap.org/api_keys).

```
